### PR TITLE
Move RefreshTask into IndexService and use since task per index

### PR DESCRIPTION
**Pull request by s1monw**
`refresh_interval` is a per index setting but we interpret and maintain it per shard. This
change moves the refresh task outside of IndexShard to the IndexService where it logically belongs
and reuses scheduling infrastructure used for translog fsync (async commit).

This change will use the same task for all shards of an index while previously we used on thread/task
per shard to refresh. This will also prevent too many concurrent refreshes if there are many indices and
shards allocated on a single node.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/IndexService.java [hunk 1]
```diff
@@ -605,40 +630,165 @@ private void maybeFSyncTranslogs() {
...
-                    indexService.maybeFSyncTranslogs();
-                    if (mustRun()) {
-                        schedule();
+        public final void run() {
+            try {
+                runInternal();
+            } catch (Exception ex) {
+                if (lastThrownException == null || sameException(lastThrownException, ex) == false) {
+                    // prevent the annoying fact of logging the same stuff all the time with an interval of 1 sec will spam all your logs
+                    indexService.logger.warn("failed to run task {} - supressing re-occuring exceptions unless the exception changes", ex, toString());
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/IndexService.java]: supressing --> suppressing

Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/IndexService.java]: re-occuring -> re-occurring

File: core/src/main/java/org/elasticsearch/index/IndexService.java [hunk 2]
```diff
@@ -605,40 +630,165 @@ private void maybeFSyncTranslogs() {
...
+            indexService.maybeFSyncTranslogs();
+        }
+
+        @Override
+        public String toString() {
+            return "translog_sync";
+        }
+    }
+
+    final class RefreshTasks extends BaseAsyncTask {
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/IndexService.java]: Maybe rename to `AysyncRefreshTask`?

File: core/src/main/java/org/elasticsearch/index/IndexService.java [hunk 3]
```diff
@@ -140,6 +145,7 @@ public IndexService(IndexSettings indexSettings, NodeEnvironment nodeEnv,
         this.listeners = new IndexingOperationListener[1+listenersIn.length];
         this.listeners[0] = slowLog;
         System.arraycopy(listenersIn, 0, this.listeners, 1, listenersIn.length);
+        this.refreshTask = new RefreshTasks(this, nodeServicesProvider.getThreadPool());
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/IndexService.java]: Can we call `getThreadPool()` in all these places that are now calling `nodeServicesProvider.getThreadPool()`?  Or maybe have a `this.threadPool` instead?

File: core/src/main/java/org/elasticsearch/index/IndexService.java [hunk 4]
```diff
@@ -310,9 +316,12 @@ public synchronized IndexShard createShard(ShardRouting routing) throws IOExcept
             eventListener.indexShardStateChanged(indexShard, null, indexShard.state(), "shard created");
             eventListener.afterIndexShardCreated(indexShard);
             indexShard.updateRoutingEntry(routing, true);
-            if (shards.isEmpty() && this.indexSettings.getTranslogSyncInterval().millis() != 0) {
+            if (shards.isEmpty()) {
                 ThreadPool threadPool = nodeServicesProvider.getThreadPool();
-                new AsyncTranslogFSync(this, threadPool).schedule(); // kick this off if we are the first shard in this service.
+                if (this.indexSettings.getTranslogSyncInterval().millis() != 0) {
+                    new AsyncTranslogFSync(this, threadPool); // kick this off if we are the first shard in this service.
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/IndexService.java]: Move this comment up just inside the `if (shards.isEmpty()) {` saying "// kick off async ops for the first shard in this index"?

File: core/src/main/java/org/elasticsearch/index/IndexService.java [hunk 5]
```diff
@@ -605,40 +630,165 @@ private void maybeFSyncTranslogs() {
...
-            if (mustRun()) {
-                threadPool.executor(ThreadPool.Names.FLUSH).execute(() -> {
-                    indexService.maybeFSyncTranslogs();
-                    if (mustRun()) {
-                        schedule();
+        public final void run() {
+            try {
+                runInternal();
+            } catch (Exception ex) {
+                if (lastThrownException == null || sameException(lastThrownException, ex) == false) {
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/IndexService.java]: This is neat ... I hope (later!) we can somehow make this generic, fold it into logger layer or something.

File: core/src/main/java/org/elasticsearch/index/IndexService.java [hunk 6]
```diff
@@ -605,40 +630,165 @@ private void maybeFSyncTranslogs() {
...
+        }
+    }
 
-    /**
-     * FSyncs the translog for all shards of this index in a defined interval.
-     */
-    final static class AsyncTranslogFSync implements Runnable {
-        private final IndexService indexService;
-        private final ThreadPool threadPool;
+    static abstract class BaseAsyncTask implements Runnable, Closeable {
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/IndexService.java]: Should this subclass `AbstractRunnable` so we are graceful about handling rejections/failures from the thread pool?

File: core/src/main/java/org/elasticsearch/index/IndexService.java [hunk 7]
```diff
@@ -310,9 +316,12 @@ public synchronized IndexShard createShard(ShardRouting routing) throws IOExcept
...
             eventListener.afterIndexShardCreated(indexShard);
             indexShard.updateRoutingEntry(routing, true);
-            if (shards.isEmpty() && this.indexSettings.getTranslogSyncInterval().millis() != 0) {
+            if (shards.isEmpty()) {
                 ThreadPool threadPool = nodeServicesProvider.getThreadPool();
-                new AsyncTranslogFSync(this, threadPool).schedule(); // kick this off if we are the first shard in this service.
+                if (this.indexSettings.getTranslogSyncInterval().millis() != 0) {
+                    new AsyncTranslogFSync(this, threadPool); // kick this off if we are the first shard in this service.
+                }
+                rescheduleRefreshTasks();
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/IndexService.java]: Hmm why do we need to do this?  The already scheduled one from the ctor should be the same?  Only if the refresh interval is changed should we need to reschedule?

Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/IndexService.java]: Oh I see, we need it so we reschedule ourselves w/ the thread pool, ok.

**Comment by mikemccand** [comment]
LGTM, it's great to push "async-ness" higher up in the stack.

File: core/src/main/java/org/elasticsearch/index/IndexService.java (hunk 6 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/IndexService.java]: I don't think scheduled threadpools supported that class yet

File: core/src/main/java/org/elasticsearch/index/IndexService.java (hunk 7 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/IndexService.java]: hmm I think I can simplify this...

**Comment by s1monw** [comment]
@mikemccand I pushed a new commit with a lot more tests and simpler task initialization.

**Comment by mikemccand** [comment]
LGTM, thanks!

**Comment by bleskes** [comment]
This is a great change. We little side effect is added debug logging each 1s / 5s. Imho this should be TRACE:

```
[2016-01-14 19:28:53,990][DEBUG][index                    ] [Sam Wilson] [index] scheduling refresh every 1s
[2016-01-14 19:28:54,995][DEBUG][index                    ] [Sam Wilson] [index] scheduling refresh every 1s
[2016-01-14 19:28:55,965][DEBUG][index                    ] [Sam Wilson] [index] scheduling translog_sync every 5s
[2016-01-14 19:28:56,000][DEBUG][index                    ] [Sam Wilson] [index] scheduling refresh every 1s
[2016-01-14 19:28:57,003][DEBUG][index                    ] [Sam Wilson] [index] scheduling refresh every 1s
[2016-01-14 19:28:58,007][DEBUG][index                    ] [Sam Wilson] [index] scheduling refresh every 1s
[2016-01-14 19:28:59,012][DEBUG][index                    ] [Sam Wilson] [index] scheduling refresh every 1s
[2016-01-14 19:29:00,017][DEBUG][index                    ] [Sam Wilson] [index] scheduling refresh every 1s
[2016-01-14 19:29:00,970][DEBUG][index                    ] [Sam Wilson] [index] scheduling translog_sync every 5s
```

I believe some of these were actually introduce in #14121 (great change), but it's hard to tell because the history is full of little commits that don't refer to relevant PR (argh  git merge). /cc @mikemccand

**Comment by mikemccand** [comment]
@bleskes I now dub you the logging policeman!

But I don't think #14121 added these debug logs?

I'm sorry for polluting git history with lots of little commits not pointing to the eventual PR ... I guess this means when I do local commits I should at least reference the ... issue I'm working against?

**Comment by bleskes** [comment]
> @bleskes I now dub you the logging policeman!

LOL. Challenge accepted. It's important :)


> I guess this means when I do local commits I should at least reference the ... issue I'm working against?

I'm a fan of squash-rebase before merging where I change the commit message to include a reference to the PR (known at that time). I really don't like how hard it has become to trace code lines to PR/Issues. I know this can start a flame war - let's please not have it here ;)

**Comment by jasontedor** [comment]
> I really don't like how hard it has become to trace code lines to PR/Issues.

I agree that it would be better if it were easier. But it can be done with a little work at the command line. Pretending that you think 0a2a7f2f7a2b19b4cbd5b954f028d9b5a9201fd9 is the culprit:

```bash
$ comm -1 -2 \
> <(git rev-list 0a2a7f2f7a2b19b4cbd5b954f028d9b5a9201fd9..master --ancestry-path) \
> <(git rev-list 0a2a7f2f7a2b19b4cbd5b954f028d9b5a9201fd9..master --first-parent) \
> | tail -1
b4a095d430f592d0a0e4f27d5beab7aed84a3441
$ git log --pretty=oneline b4a095d430f592d0a0e4f27d5beab7aed84a3441 | head -1
b4a095d430f592d0a0e4f27d5beab7aed84a3441 Merge pull request #14121 from mikemccand/fair_indexing_buffers
```

Additionally, the [GitHub search](https://github.com/search?q=0a2a7f2f7a2b19b4cbd5b954f028d9b5a9201fd9&type=Issues&utf8=✓) usually gets it there for me, as well (the key is to search issues).

**Comment by s1monw** [comment]
@bleskes FYI https://github.com/elastic/elasticsearch/commit/903d0ff96ce662ddf4f2d6d27a753c6ab8a9fbeb

File: core/src/main/java/org/elasticsearch/index/IndexService.java [hunk 8]
```diff
@@ -605,40 +627,170 @@ private void maybeFSyncTranslogs() {
...
+                }
+            } finally {
+                onTaskCompletion();
+            }
+        }
+
+        private static boolean sameException(Exception left, Exception right) {
+            if (left.getClass() == right.getClass()) {
+                if ((left.getMessage() != null && left.getMessage().equals(right.getMessage()))
+                    || left.getMessage() == right.getMessage()) {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/IndexService.java]: Cleaner is `Objects.equals(left.getMessage(), right.getMessage())`.

**Comment by bleskes** [comment]
@jasontedor thanks for git voodoo. I'll use it if need be. Personally - I just want git blame (and it's integration with IntelliJ ) to work

@s1monw re:
> @bleskes FYI 903d0ff

Thank you!

**Pull request merged**